### ISSUE TITLE
Making everything ECS would render ViewVariables useless.

### ISSUE THREAD

**Issue by gradientvera** [space-wizards/RobustToolbox#1772]
This is easier to explain with an example.
Before StackComponent in SS14 was stripped of all logic, it had a `Count` property with a setter that would `Dirty()` the component, change appearance, raise an event, etc.
After I refactored it, `Count` is a simple auto-property with no special logic in the setter. You need to raise a `StackChangeCount` event in order to change this property, and an EntitySystem will take care of all the logic `Count`'s setter did before.
However, now editing `Count` in VV is not ideal. It doesn't perform any of the logic it should, and there's no current way to fix this.
I am not sure how this should be handled, or what should be done in order to ensure VV doesn't become useless after we transition to a more ECS approach of coding content.

**Comment by RemieRichards** [comment]
It'd be *half* useless, given *VIEW* Variables is also for viewing.

Might be fine to just add appropriate properties specifically for VV, which do have logic in them (this technically violates the ECS nature, but given that the actual main flow of the code would still *not* be going through them, it'd be fine, no?)

**Comment by gradientvera** [comment]
That is a solution indeed, but I don't particularly like it due to its hacky nature, hmm... If nobody can come up with a better solution, I think we'll just have to do that.

**Comment by Acruid** [comment]
Another solution is to use CSI instead of VV.

**Comment by gradientvera** [comment]
> Another solution is to use CSI instead of VV.

Sadly this is not an option, as we can't give every admin SCSI perms

**Comment by RemieRichards** [comment]
CSI is way too dangerous to give to the average admin

**Comment by PaulRitter** [comment]
i guess the easiest way would be to make the entitysystem somehow subscribe to changes to the autoproperty. am thinking emitting some funny il into the component, then having a generic method on the entitysystem that can just subscribe to changes to a specific field. maybe even just emit the funny il if a function-call for the field in question exists?

imo this would also be a function specific to VV, so you could just go "every field that has the system subscribe to changes can be edited, the others are readonly"?

**Comment by gradientvera** [comment]
Having to subscribe to every single property that can be changed by VV on entity systems seems far from ideal, but I can see why.
One problem is that the EntitySystems themselves change the property when they get the event, so you have to somehow distinguish between something like VV/animatable changing it and the entity system changing it when handling the event itself.

**Comment by gradientvera** [comment]
So I had a discussion with Paul on discord today and he proposed a [great idea](https://discord.com/channels/310555209753690112/770682801607278632/852446307981328384).
All of the problems here come from things changing object data using *reflection*, so we should add APIs to `IReflectionManager` to change fields/properties!

First, we would add an attribute which specifies an event type.
The event type should inherit `EntityEventArgs`, have a constructor with no params, or a constructor with a single param of the same type as the field/property.

Second, we would add the APIs to the reflection manager. They would allow you to change a field/property of any object, and also take in both an EntityUid to raise a directed event when applicable and a bool for whether to raise the event broadcasted too or not.

Third, we would change things using reflection to this API and boom, problem solved!

What do you people think? If this is fine with everyone, I can implement it.

**Comment by PaulRitter** [comment]
i obviously think this is a good solution. very contained and straight-forward to use

**Comment by gradientvera** [comment]
> i obviously think this is a good solution. very contained and straight-forward to use

PJB rejected it. Honestly, after I implemented it, it just felt like an awful hack... And we're moving away from method events in favor of public methods in Entity Systems, so the solution is no longer applicable

**Comment by gradientvera** [comment]
The simplest solution I can personally think of (without rewriting VV and animations) is manually writing custom properties for everything that may be edited with VV or animated with the animation player that would look like this:
```csharp
        // Property for VV and animations
        [ViewVariables(VVAccess.ReadWrite)]
        public int SetCount
        {
            get => Count;
            set => EntitySystem.Get<SharedStackSystem>().SetCount(Owner.Uid, this, value);
        }
        
        /// <summary>
        ///     Actual current stack count.
        ///     Do NOT set this directly, use the <see cref="SharedStackSystem"/> instead.
        /// </summary>     
        [DataField("count")]
        public int Count { get; set; } = 30;
```

**Comment by SweptWasTaken** [comment]
> 
> 
> The simplest solution I can personally think of (without rewriting VV and animations) is manually writing custom properties for everything that may be edited with VV or animated with the animation player that would look like this:
> 
> ```cs
>         // Property for VV and animations
>         [ViewVariables(VVAccess.ReadWrite)]
>         public int SetCount
>         {
>             get => Count;
>             set => EntitySystem.Get<SharedStackSystem>().SetCount(Owner.Uid, this, value);
>         }
>         
>         /// <summary>
>         ///     Actual current stack count.
>         ///     Do NOT set this directly, use the <see cref="SharedStackSystem"/> instead.
>         /// </summary>     
>         [DataField("count")]
>         public int Count { get; set; } = 30;
> ```

Won't this inflate the beginning of .cs files?

**Comment by gradientvera** [comment]
> Won't this inflate the beginning of .cs files?

Indeed, this would be a lot of boilerplate for something that could be automatized/solved in a different way, which is why I haven't implemented this in stacks. I wanna see if we can come up with a better solution.

**Comment by SweptWasTaken** [comment]
> 
> 
> > Won't this inflate the beginning of .cs files?
> 
> Indeed, this would be a lot of boilerplate for something that could be automatized/solved in a different way, which is why I haven't implemented this in stacks. I wanna see if we can come up with a better solution.

View variables are gonna be much more admin facing than anything else in the codebase. Is there anyway to write them in separate yaml files? 

Just spit-balling I have little knowledge on how this stuff connects together yet.

**Comment by gradientvera** [comment]
> View variables are gonna be much more admin facing than anything else in the codebase. Is there anyway to write them in separate yaml files?

No, not really viable or something we want

**Comment by gradientvera** [comment]
Fixed by #3214
